tests/userRegistration.test.ts

```diff
@@ -0,0 +1,115 @@
+import request from "supertest";
+import app from "../src/server";
+import prisma from "../src/utils/prisma";
+
+describe("User Registration", () => {
+  describe("Successful registration", () => {
+    afterAll(async () => {
+      await prisma.user.deleteMany({
+        where: {
+          email: {
+            contains: "test.com",
+          },
+        },
+      });
+    });
+
+    test("User can be created with only email and password provided", async () => {
+      const res = await request(app).post("/api/register").send({
+        email: "test1@test.com",
+        password: "Test123Pass!",
+      });
+
+      expect(res.statusCode).toBe(201);
+      expect(res.body).toEqual({
+        status: "success",
+        message: "User created successfully",
+      });
+    });
+
+    test("User can be created with all details provided", async () => {
+      const res = await request(app).post("/api/register").send({
+        firstName: "John",
+        lastName: "Doe",
+        phoneNumber: 1555123456,
+        address: "123 Main Street, New York, USA",
+        email: "test2@test.com",
+        password: "Test123Pass!",
+      });
+
+      expect(res.statusCode).toBe(201);
+      expect(res.body).toEqual({
+        status: "success",
+        message: "User created successfully",
+      });
+    });
+  });
+
+  describe("Unsuccessful registration", () => {
+    test("User can't use already taken email address", async () => {
+      await request(app).post("/api/register").send({
+        email: "test@test.com",
+        password: "Test123Pass!",
+      });
+
+      const res = await request(app).post("/api/register").send({
+        email: "test@test.com",
+        password: "Test123Pass!",
+      });
+
+      expect(res.statusCode).toBe(409);
+
+      await prisma.user.delete({
```

Comment:
Have you tried to use `beforeEach` to delete all users in database so that you don't have to delete it in the test after your assertions?

Reply:
At the moment, we are using the main database on Supabase, i.e. the production one. We are not sure if we should create a new one, just for testing purposes. 

That's why I used `afterEach`, to delete the users created during the tests when tests are complete.

Reply:
Hmm, I double checked your test. Since you are not testing database here, you are testing that you endpoint is returning the right data I would now call real database.
Find the way to most prisma client.

And you can also chain `supertest` requests. Something like:

```
equest(app).post("/api/register").send({
        email: "test1@test.com",
        password: "Test123Pass!",
      }).expect(...);
```

Reply:
I'm trying to make a test `User can't use already taken email address` pass for the past few days, but without success. I've mocked the Prisma client. I've found an explanation how to mock HTTP requests and responses using Nock library, but I didn't have success with it. I mocked the response to return 409, but test always shows that 201 was returned.

Any suggestions what could be wrong? The file in question is `userRegistration.test.ts`, lines 56-68.

Reply:
You started in the right way, since you followed the prisma docs to create `singleton.ts`. Your mistake was that you didn't understand nock, `nock` is for mocking REST calls, and you don't have any REST call in your `user.route.ts`. What you have there is `user.service` and prisma client in it. So you need to mock prisma client, which you already made with `singleton.ts`. All I had to do is to mock `findUnique()` and thest is green.

Reply:
Now please use this test as an example and fix all other tests so that you don't call real database but to use `prismaMock`.